### PR TITLE
add NE <> OSM data join for places layer

### PR DESCRIPTION
**Pull request by nvkelso**
Fixes https://github.com/protomaps/basemaps/issues/91.

### CODE REVIEW

File: tiles/src/main/java/com/protomaps/basemap/layers/Places.java [hunk 1]
```diff
@@ -25,6 +29,17 @@ public String name() {
 
   private final AtomicInteger placeNumber = new AtomicInteger(0);
 
+  // To have consistent min_zoom and other attr between Natural Earth (low zooms) and OpenStreetMap (high zooms)
+  // we need to store the NE places with a spatial indexes for later joins with OSM
+  private PointIndex<NaturalEarthPlace> ne_populated_places = PointIndex.create();
```
Comment (nvkelso) [inline tiles/src/main/java/com/protomaps/basemap/layers/Places.java:34]: After creation inserts are blocked?

File: tiles/src/main/java/com/protomaps/basemap/layers/Places.java [hunk 2]
```diff
@@ -52,6 +72,22 @@ public void processNe(SourceFeature sf, FeatureCollector features) {
     if (!sourceLayer.equals("ne_10m_populated_places")) {
       return;
     }
+
+    // Setup for high zoom content
```
Comment (nvkelso) [inline tiles/src/main/java/com/protomaps/basemap/layers/Places.java:76]: Can this be moved out of per tile processing into project startup?

Comment (bdon) [reply tiles/src/main/java/com/protomaps/basemap/layers/Places.java:76]: I think it's a good idea, you said only 8000 places right? I don't want to slow down the small job tests by loading in the whole thing every time

File: tiles/src/main/java/com/protomaps/basemap/layers/Places.java [hunk 3]
```diff
@@ -25,6 +29,17 @@ public String name() {
...
 
+  // To have consistent min_zoom and other attr between Natural Earth (low zooms) and OpenStreetMap (high zooms)
+  // we need to store the NE places with a spatial indexes for later joins with OSM
+  private PointIndex<NaturalEarthPlace> ne_populated_places = PointIndex.create();
+
+  // Data structure for any Natural Earth place (eg for data joins between NE and OSM)
+  private record NaturalEarthPlace(String name, String wikidataId, float minZoom, int populationRank) {}
+
+  // Search window size for NE <> OSM data joins
+  private static final double LOCALITY_JOIN_DISTANCE = GeoUtils.metersToPixelAtEquator(0, 50_000) / 256d;
```
Comment (bdon) [inline tiles/src/main/java/com/protomaps/basemap/layers/Places.java:40]: from what I read below, this is only necessary in 2% of the cases that don't have a Wikidata ID?

File: tiles/src/main/java/com/protomaps/basemap/layers/Places.java (hunk 1 above)
Comment (nvkelso) [reply tiles/src/main/java/com/protomaps/basemap/layers/Places.java:34]: Verified this works in a build with no errors... as long as NE is added to Planetiler profile before OSM.

File: tiles/src/main/java/com/protomaps/basemap/layers/Places.java (hunk 2 above)
Comment (nvkelso) [reply tiles/src/main/java/com/protomaps/basemap/layers/Places.java:76]: The current config doesn't seem to noticeably affect build time so it's probably OK as is?

But it's better architecture to only load it once on project init. We can file a followup issue for that?

**Comment by nvkelso** [comment]
Before low zoom:

image

Before high zoom:

image

After high zoom:

image

**Comment by nvkelso** [comment]
@bdon back at ya

File: tiles/src/main/java/com/protomaps/basemap/layers/Places.java (hunk 2 above)
Comment (nvkelso) [reply tiles/src/main/java/com/protomaps/basemap/layers/Places.java:90]: Followup in https://github.com/protomaps/basemaps/issues/108.

**Pull request merged**